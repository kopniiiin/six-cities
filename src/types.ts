export enum OfferType {
  APARTMENT = `apartment`,
  HOTEL = `hotel`,
  HOUSE = `house`,
  ROOM = `room`
}

export enum City {
  PARIS = `Paris`,
  COLOGNE = `Cologne`,
  BRUSSELS = `Brussels`,
  AMSTERDAM = `Amsterdam`,
  HAMBURG = `Hamburg`,
  DUSSELDORF = `Dusseldorf`
}

export enum SortType {
  POPULAR = `popular`,
  TO_HIGH_PRICE = `price: low to high`,
  TO_LOW_PRICE = `price: high to low`,
  TOP_RATED = `top rated first`
}

export enum Path {
  MAIN = `/`,
  LOGIN = `/login`,
  OFFER = `/offer`,
  FAVORITES = `/favorites`
}

export enum ServerURL {
  LOGIN = `/login`,
  OFFERS = `/hotels`,
  FAVORITES = `/favorite`,
  REVIEWS = `/comments`
}

export enum ServerResponseStatus {
  OK = 200,
  UNAUTHORIZED = 401
}

export enum AuthorizationStatus {
  AUTHORIZED = `AUTHORIZED`,
  UNAUTHORIZED = `UNAUTHORIZED`
}

export interface Id {
  id: string;
}

export interface Location {
  coordinates: number[];
}

export interface User {
  name: string;
  photo: string;
}

export type Offer = Id & {
  type: OfferType;
  name: string;
  mainPhoto: string;
  isFavorite: boolean;
  isPremium: boolean;
  rating: number;
  price: number;
}

export interface Review {
  date: string;
  text: string;
  rating: number;
  user: User;
}

export type ReviewWithId = Id & Review;

export interface ReviewData {
  text: string;
  rating: number;
}

export interface AuthorizationData {
  email: string;
  password: string;
}
