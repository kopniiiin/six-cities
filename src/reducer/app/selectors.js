import NameSpace from "../name-space";

export const getActiveCity = (state) => state[NameSpace.APP].activeCity;

export const getActiveSortType = (state) => state[NameSpace.APP].activeSortType;
