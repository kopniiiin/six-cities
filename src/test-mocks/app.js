import {OfferType, City, SortType, AuthorizationStatus} from "../const.js";

export default {
  authorizationStatus: AuthorizationStatus.UNAUTHORIZED,
  activeCity: City.AMSTERDAM,
  activeSortType: SortType.POPULAR,
  offers: [{
    id: `4`,
    type: OfferType.ROOM,
    name: `Paper place`,
    mainPhoto: `photo`,
    isFavorite: false,
    isPremium: false,
    rating: 4,
    price: 4,
    location: {coordinates: [4, 4]},
    city: {location: {coordinates: [4, 4], zoom: 4}}
  }],
  onCityClick: () => {},
  onSortTypeChange: () => {},
  onLoginScreenSubmit: () => {},
  onOfferFavoritenessChange: () => {}
};
