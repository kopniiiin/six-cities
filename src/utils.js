const MONTHS = [
  `January`,
  `February`,
  `March`,
  `April`,
  `May`,
  `June`,
  `July`,
  `August`,
  `September`,
  `October`,
  `November`,
  `December`
];

export const extend = (target, update) => Object.assign({}, target, update);

export const pushElement = (array, element) => [...array, element];

export const replaceElement = (array, element, index) => [...array.slice(0, index), element, ...array.slice(index + 1)];

export const removeElement = (array, index) => [...array.slice(0, index), ...array.slice(index + 1)];

export const upperCaseFirstLetter = (string) => string[0].toUpperCase() + string.slice(1);

export const formatDate = (dateString) => {
  const date = new Date(dateString);
  return `${MONTHS[date.getMonth()]} ${date.getDate()}, ${date.getFullYear()}`;
};
