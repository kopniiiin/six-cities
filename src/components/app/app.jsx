import React, {PureComponent} from "react";
import {BrowserRouter, Switch, Route} from "react-router-dom";
import {connect} from "react-redux";
import PropTypes from "prop-types";

import {City, SortType, Path, AuthorizationStatus} from "../../const.js";

import GuestRoute from "../guest-route/guest-route.jsx";
import Header from "../header/header.jsx";
import Main from "../main/main.jsx";
import LoginScreen from "../login-screen/login-screen.jsx";
import OfferScreen from "../offer-screen/offer-screen.jsx";

import withActiveItem from "../../hocs/with-active-item/with-active-item.jsx";
import withAuthorizationData from "../../hocs/with-authorization-data/with-authorization-data.jsx";

import {ActionCreator as AppActionCreator} from "../../reducer/app/app.js";
import {getActiveCity, getActiveSortType} from "../../reducer/app/selectors.js";
import {Operation as UserOperation} from "../../reducer/user/user.js";
import {getAuthorizationStatus, getEmail} from "../../reducer/user/selectors.js";
import {Operation as OffersOperation} from "../../reducer/offers/offers.js";
import {getFilteredAndSortedOffers} from "../../reducer/offers/selectors.js";

const MainWithActiveItem = withActiveItem(Main);
const LoginScreenWithAuthorizationData = withAuthorizationData(LoginScreen);

const Screen = {
  MAIN: `MAIN`,
  LOGIN: `LOGIN`,
  OFFER: `OFFER`
};

const offerScreenPropTypesCopy = Object.assign({}, OfferScreen.propTypes);
offerScreenPropTypesCopy.id = PropTypes.string.isRequired;
delete offerScreenPropTypesCopy.nearOffers;

const propTypes = {
  authorizationStatus: PropTypes.oneOf(Object.values(AuthorizationStatus)).isRequired,
  email: PropTypes.string,
  activeCity: PropTypes.oneOf(Object.values(City)).isRequired,
  activeSortType: PropTypes.oneOf(Object.values(SortType)).isRequired,
  offers: PropTypes.arrayOf(PropTypes.shape(offerScreenPropTypesCopy)).isRequired,
  onCityClick: PropTypes.func.isRequired,
  onSortTypeChange: PropTypes.func.isRequired,
  onLoginScreenSubmit: PropTypes.func.isRequired,
  onOfferFavoritenessChange: PropTypes.func.isRequired
};

class App extends PureComponent {
  constructor(props) {
    super(props);
    this._handleLoginScreenSubmit = this._handleLoginScreenSubmit.bind(this);
    this._handleOfferCardBookmarkButtonClick = this._handleOfferCardBookmarkButtonClick.bind(this);
  }

  render() {
    const {
      authorizationStatus,
      email,
      activeCity,
      activeSortType,
      offers,
      onCityClick,
      onSortTypeChange,
    } = this.props;

    const header = <Header email={email}/>;

    return (
      <BrowserRouter>
        <Switch>

          <Route path={Path.MAIN} exact render={() => (
            <MainWithActiveItem
              activeCity={activeCity}
              activeSortType={activeSortType}
              offers={offers}
              onCityClick={onCityClick}
              onSortTypeChange={onSortTypeChange}
              onOfferCardBookmarkButtonClick={this._handleOfferCardBookmarkButtonClick}>
              {header}
            </MainWithActiveItem>
          )}/>

          <GuestRoute authorizationStatus={authorizationStatus} path={Path.LOGIN} exact render={() => (
            <LoginScreenWithAuthorizationData
              activeCity={activeCity}
              onSubmit={this._handleLoginScreenSubmit}>
              {header}
            </LoginScreenWithAuthorizationData>
          )}/>

          <Route path={`${Path.OFFER}/:offerId`} exact render={({match: {params: {offerId}}}) => (
            <OfferScreen
              authorizationStatus={authorizationStatus}
              id={offerId}
              onOfferCardBookmarkButtonClick={this._handleOfferCardBookmarkButtonClick}>
              {header}
            </OfferScreen>
          )}/>

        </Switch>
      </BrowserRouter>
    );
  }

  _handleLoginScreenSubmit(authorizationData) {
    const {onLoginScreenSubmit} = this.props;

    onLoginScreenSubmit(authorizationData);
  }

  _handleOfferCardBookmarkButtonClick(offerId) {
    const {authorizationStatus, onOfferFavoritenessChange} = this.props;

    if (authorizationStatus === AuthorizationStatus.AUTHORIZED) {
      onOfferFavoritenessChange(offerId);
    }
  }
}

App.propTypes = propTypes;

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  email: getEmail(state),
  activeCity: getActiveCity(state),
  activeSortType: getActiveSortType(state),
  offers: getFilteredAndSortedOffers(state)
});

const mapDispatchToProps = (dispatch) => ({
  onCityClick: (city) => dispatch(AppActionCreator.setActiveCity(city)),
  onSortTypeChange: (sortType) => dispatch(AppActionCreator.setActiveSortType(sortType)),
  onLoginScreenSubmit: (authorizationData) => dispatch(UserOperation.login(authorizationData)),
  onOfferFavoritenessChange: (offerId) => dispatch(OffersOperation.toggleFavoriteness(offerId))
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
