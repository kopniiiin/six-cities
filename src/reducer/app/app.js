import {DEFAULT_CITY, DEFAULT_SORT_TYPE} from "../../const";

import {extend} from "../../utils";

const initialState = {
  activeCity: DEFAULT_CITY,
  activeSortType: DEFAULT_SORT_TYPE
};

export const ActionType = {
  SET_ACTIVE_CITY: `SET_ACTIVE_CITY`,
  SET_ACTIVE_SORT_TYPE: `SET_ACTIVE_SORT_TYPE`
};

export const ActionCreator = {
  setActiveCity: (activeCity) => ({type: ActionType.SET_ACTIVE_CITY, payload: activeCity}),
  setActiveSortType: (activeSortType) => ({type: ActionType.SET_ACTIVE_SORT_TYPE, payload: activeSortType})
};

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_ACTIVE_CITY:
      return extend(state, {activeCity: action.payload});
    case ActionType.SET_ACTIVE_SORT_TYPE:
      return extend(state, {activeSortType: action.payload});
  }

  return state;
};
