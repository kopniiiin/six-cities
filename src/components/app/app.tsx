import React, {PureComponent} from "react";
import {Router, Switch, Route} from "react-router-dom";
import {connect} from "react-redux";
import PropTypes from "prop-types";

import history from "../../history";

import {OfferType, City, SortType, Path, AuthorizationStatus} from "../../const";

import GuestRoute from "../guest-route/guest-route";
import PrivateRoute from "../private-route/private-route";
import ErrorMessage from "../error-message/error-message";
import Header from "../header/header";
import Main from "../main/main";
import LoginScreen from "../login-screen/login-screen";
import OfferScreen from "../offer-screen/offer-screen";
import FavoritesScreen from "../favorites-screen/favorites-screen";

import withActiveItem from "../../hocs/with-active-item/with-active-item";
import withAuthorizationData from "../../hocs/with-authorization-data/with-authorization-data";

import {ActionCreator as AppActionCreator} from "../../reducer/app/app";
import {getActiveCity, getActiveSortType} from "../../reducer/app/selectors";
import {Operation as UserOperation} from "../../reducer/user/user";
import {getAuthorizationStatus, getEmail} from "../../reducer/user/selectors";
import {Operation as OffersOperation} from "../../reducer/offers/offers";
import {getError, getFilteredAndSortedOffers} from "../../reducer/offers/selectors";

const MainWithActiveItem = withActiveItem(Main);
const LoginScreenWithAuthorizationData = withAuthorizationData(LoginScreen);

const offerScreenPropTypesCopy = Object.assign({}, OfferScreen.propTypes);
offerScreenPropTypesCopy.id = PropTypes.string.isRequired;
delete offerScreenPropTypesCopy.nearOffers;

const propTypes = {
  authorizationStatus: PropTypes.oneOf(Object.values(AuthorizationStatus)).isRequired,
  email: PropTypes.string,
  activeCity: PropTypes.oneOf(Object.values(City)).isRequired,
  activeSortType: PropTypes.oneOf(Object.values(SortType)).isRequired,
  error: PropTypes.string,
  offers: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    type: PropTypes.oneOf(Object.values(OfferType)).isRequired,
    name: PropTypes.string.isRequired,
    mainPhoto: PropTypes.string.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    isPremium: PropTypes.bool.isRequired,
    rating: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    location: PropTypes.shape({
      coordinates: PropTypes.arrayOf(PropTypes.number).isRequired
    }).isRequired,
    city: PropTypes.shape({
      location: PropTypes.shape({
        coordinates: PropTypes.arrayOf(PropTypes.number).isRequired,
        zoom: PropTypes.number.isRequired
      }).isRequired
    }).isRequired
  })).isRequired,
  onCityClick: PropTypes.func.isRequired,
  onSortTypeChange: PropTypes.func.isRequired,
  onLoginScreenSubmit: PropTypes.func.isRequired,
  onOfferFavoritenessChange: PropTypes.func.isRequired
};

class App extends PureComponent {
  constructor(props) {
    super(props);
    this._handleLoginScreenSubmit = this._handleLoginScreenSubmit.bind(this);
    this._handleOfferCardBookmarkButtonClick = this._handleOfferCardBookmarkButtonClick.bind(this);
  }

  render() {
    const {
      authorizationStatus,
      email,
      activeCity,
      activeSortType,
      error,
      offers,
      onCityClick,
      onSortTypeChange,
    } = this.props;

    const header = <Header email={email}>{error && <ErrorMessage text={error}/>}</Header>;

    return (
      <Router history={history}>
        <Switch>

          <Route path={Path.MAIN} exact render={() => (
            <MainWithActiveItem
              activeCity={activeCity}
              activeSortType={activeSortType}
              offers={offers}
              onCityClick={onCityClick}
              onSortTypeChange={onSortTypeChange}
              onOfferCardBookmarkButtonClick={this._handleOfferCardBookmarkButtonClick}>
              {header}
            </MainWithActiveItem>
          )}/>

          <GuestRoute authorizationStatus={authorizationStatus} path={Path.LOGIN} exact render={() => (
            <LoginScreenWithAuthorizationData
              activeCity={activeCity}
              onSubmit={this._handleLoginScreenSubmit}>
              {header}
            </LoginScreenWithAuthorizationData>
          )}/>

          <Route path={`${Path.OFFER}/:offerId`} exact render={({match: {params: {offerId}}}) => (
            <OfferScreen
              authorizationStatus={authorizationStatus}
              id={offerId}
              onOfferCardBookmarkButtonClick={this._handleOfferCardBookmarkButtonClick}>
              {header}
            </OfferScreen>
          )}/>

          <PrivateRoute authorizationStatus={authorizationStatus} path={Path.FAVORITES} exact render={() => (
            <FavoritesScreen
              onOfferCardBookmarkButtonClick={this._handleOfferCardBookmarkButtonClick}>
              {header}
            </FavoritesScreen>
          )}/>

        </Switch>
      </Router>
    );
  }

  _handleLoginScreenSubmit(authorizationData) {
    const {onLoginScreenSubmit} = this.props;

    onLoginScreenSubmit(authorizationData);
  }

  _handleOfferCardBookmarkButtonClick(offerId) {
    const {authorizationStatus, onOfferFavoritenessChange} = this.props;

    if (authorizationStatus === AuthorizationStatus.AUTHORIZED) {
      onOfferFavoritenessChange(offerId);
      return;
    }

    history.push(Path.LOGIN);
  }
}

App.propTypes = propTypes;

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  email: getEmail(state),
  activeCity: getActiveCity(state),
  activeSortType: getActiveSortType(state),
  error: getError(state),
  offers: getFilteredAndSortedOffers(state)
});

const mapDispatchToProps = (dispatch) => ({
  onCityClick: (city) => dispatch(AppActionCreator.setActiveCity(city)),
  onSortTypeChange: (sortType) => dispatch(AppActionCreator.setActiveSortType(sortType)),
  onLoginScreenSubmit: (authorizationData) => dispatch(UserOperation.login(authorizationData)),
  onOfferFavoritenessChange: (offerId) => dispatch(OffersOperation.toggleFavoriteness(offerId))
});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
