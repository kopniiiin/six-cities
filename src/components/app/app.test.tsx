import * as React from "react";
import {configure, shallow} from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

import {OfferType, City, SortType, AuthorizationStatus} from "../../types";

import {doNothing} from "../../utils";

import {App} from "./app";

configure({adapter: new Adapter()});

const testMocks = {
  authorizationStatus: AuthorizationStatus.UNAUTHORIZED,
  activeCity: City.AMSTERDAM,
  activeSortType: SortType.POPULAR,
  offers: [{
    id: `4`,
    type: OfferType.ROOM,
    name: `Paper place`,
    mainPhoto: `photo`,
    isFavorite: false,
    isPremium: false,
    rating: 4,
    price: 4,
    location: {coordinates: [4, 4]},
    city: {location: {coordinates: [4, 4], zoom: 4}}
  }],
  onCityClick: doNothing,
  onSortTypeChange: doNothing,
  onLoginScreenSubmit: doNothing,
  onOfferFavoritenessChange: doNothing
};

describe(`snapshot test: App component`, () => {
  it(`should render correctly`, () => expect(
      toJson(shallow(<App {...testMocks}/>))
  ).toMatchSnapshot());
});
