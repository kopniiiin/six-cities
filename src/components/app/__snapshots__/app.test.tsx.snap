// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: App component should render correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Switch>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <GuestRoute
      authorizationStatus="UNAUTHORIZED"
      exact={true}
      path="/login"
      render={[Function]}
    />
    <Route
      exact={true}
      path="/offer/:offerId"
      render={[Function]}
    />
    <PrivateRoute
      authorizationStatus="UNAUTHORIZED"
      exact={true}
      path="/favorites"
      render={[Function]}
    />
  </Switch>
</Router>
`;
